File: Fermat.c
#include<stdio.h>
#include<math.h>
int main()
{
    int M;
    scanf("%d",&M);
    int arr[M],front=0,rear=1;
    arr[0]=M;
    while (front!=rear)
    {
        int n = arr[front];
        int a = ceil(sqrt(n));

        while(1)
        {
            int b2 = a*a - n;
            int b = sqrt(b2);
            if(b*b==b2)
            {
                int f1=a+b,f2=a-b;
                if(f1==1)
                {
                    printf("%d ",f2);
            
                }
                else if(f2==1)
                {
                    printf("%d ",f1);
                    
                   
                }
                else
                {
                    arr[rear++] = f1;
                    arr[rear++] = f2;
                }
                front++;
                break;
            }
            else
            {
                a++;
            }
        }
       
    }
    
}

File: Fermat.java
import java.util.*;
public class Fermat {
    public static void main(String[] args) {
            int N;
            Scanner sc = new Scanner(System.in);
            N = sc.nextInt();
            int arr[] = new int[N];
            int prime[] = new int[N];
            int g=0;
            arr[0] = N;
            int front=0,rear=1;
            while(front!=rear)
            {
            int s = arr[front];
            int a = (int)Math.ceil(Math.sqrt(s));
            while(true)
            {
                int b2 = a*a - s;
                int b = (int)Math.sqrt(b2);
                if(b*b==b2)
                {
                    if((a+b)==1 || (a-b)==1)
                    {
                        prime[g++] = s;
                        front++;
                        break;
                    }
                    arr[rear++] = a+b;
                    arr[rear++] = a-b;
                    front++;
                    break;
                }
                else
                {
                    a++;
                }
            }
        }
        for(int i=0;i<g;i++)
        {
            System.out.print(prime[i]+ " ");
        }
    }
}
